Банки хранят огромные объёмы информации о своих клиентах. Эти данные можно использовать для того, чтобы оставаться на связи с клиентами и индивидуально ориентировать их на подходящие именно им продукты или банковские предложения.

Обычно с выбранными клиентами связываются напрямую через разные каналы связи: лично (например, при визите в банк), по телефону, по электронной почте, в мессенджерах и так далее. Этот вид маркетинга называется прямым маркетингом. На самом деле, прямой маркетинг используется для взаимодействия с клиентами в большинстве банков и страховых компаний. Но, разумеется, проведение маркетинговых кампаний и взаимодействие с клиентами — это трудозатратно и дорого.

Банкам хотелось бы уметь выбирать среди своих клиентов именно тех, которые с наибольшей вероятностью воспользуются тем или иным предложением, и связываться именно с ними.

Вам предоставили данные о последней маркетинговой кампании, которую проводил банк: задачей было привлечь клиентов для открытия депозита. Вы должны проанализировать эти данные, выявить закономерность и найти решающие факторы, повлиявшие на то, что клиент вложил деньги именно в этот банк. Если вы сможете это сделать, то поднимете доходы банка и поможете понять целевую аудиторию, которую необходимо привлекать путём рекламы и различных предложений.

Данные принадлежат реальному банку, и решение этого кейса станет хорошим образцом для вашего портфолио.

Бизнес-задача: определить характеристики, по которым можно выявить клиентов, более склонных к открытию депозита в банке, и за счёт этого повысить результативность маркетинговой кампании.

Техническая задача для вас как для специалиста в Data Science: построить модель машинного обучения, которая на основе предложенных характеристик клиента будет предсказывать, воспользуется он предложением об открытии депозита или нет.

Ваши основные цели:

Исследуйте данные, а не просто вычисляйте метрики и создавайте визуализации.
Попробуйте выявить характерные черты для потенциальных клиентов, чтобы чётко очертить ЦА и увеличить прибыль банка.
Проявляйте фантазию и используйте разные инструменты для повышения качества прогноза.
Организационная информация

Проект будет состоять из пяти частей:

1
Первичная обработка данных

В рамках этой части вам предстоит обработать пропуски и выбросы в данных. Это необходимо для дальнейшей работы с ними.

2
Разведывательный анализ данных (EDA)

Вам необходимо будет исследовать данные, нащупать первые закономерности и выдвинуть гипотезы.

3
Отбор и преобразование признаков

На этом этапе вы перекодируете и преобразуете данные таким образом, чтобы их можно было использовать при решении задачи классификации. Если на первом этапе вы лишь избавите данные от ненужных артефактов, то на этом шаге совершите действия, более важные для подготовки данных к задаче классификации, уже понимая их структуру.

4
Решение задачи классификации: логистическая регрессия и решающие деревья

На данном этапе вы построите свою первую прогностическую модель и оцените её качество. Вы научитесь подбирать оптимальные параметры модели для того, чтобы получить наилучший результат для конкретного алгоритма.

5
Решение задачи классификации: ансамбли моделей и построение прогноза

На заключительном этапе вы сможете доработать своё предсказание с использованием более сложных алгоритмов и оценить, с помощью какой модели возможно сделать более качественные прогнозы.

Каждая из представленных частей проекта будет состоять из блока практических заданий, которые вам необходимо выполнить в своих Jupyter-ноутбуках, и контрольных вопросов на платформе, отвечая на которые вы сможете проверять верность своего решения.

!

Отправлять итоговый ноутбук на проверку ментору не требуется. Возникающие вопросы вы можете задать ментору в соответствующем канале в Telegram, а также на групповой консультации по проекту.

Выполняйте задания строго последовательно: при нарушении последовательности выполнения заданий проверка ответов не сможет корректно оценить ваши решения.

Важно, что иногда задания не несут практической ценности (т. е. какие-то из показателей не нужно было бы рассчитывать при решении задачи в реальной ситуации), однако они отлично помогают идентифицировать правильность выполнения вами необходимых действий.

Что необходимо сделать для успешного выполнения проекта?

Скачайте датасет и ноутбук-шаблон. Проверьте, что файлы скачались в корректном формате: датасет должен быть в формате CSV, а ноутбук — в формате .IPYNB.
Внимательно изучите детали задачи и данные. Данные реальные, без предварительной обработки. Уделите особое внимание предобработке данных, учитывайте все нюансы.
Пользуйтесь советами и подсказками, приведёнными по ходу выполнения проекта, но не злоупотребляйте ими — сначала попытайтесь выполнить задания самостоятельно. Если вы не знаете, как решить то или иное задание из-за пробелов в знаниях по Python, не забывайте обращаться к документации, материалам курса, поиску в интернете. Представьте, что это реальный кейс, и необходимо найти всю информацию самостоятельно. Однако, если это вызовет затруднения, вы всегда можете поэтапно пользоваться подсказками.
Ответьте на все контрольные вопросы. Внимательно читайте вопросы и не забывайте про жёсткие требования к их последовательности: любое нарушение порядка действий может повлечь невозможность получить верные ответы.
Загрузите ноутбук со своим решением на GitHub, аккуратно его оформив. Несмотря на то, что ваши ответы не будут проверяться ментором, решение реального кейса может стать хорошим вкладом в ваше портфолио.
Рекомендации по оформлению ноутбука-решения

Оформите решение в Jupyter Notebook. После выполнения задания не забудьте сохранить результат в корректном формате .IPYNB.
Возьмите за основу ноутбук-шаблон: не следует менять последовательность действий и уже заполненные ячейки (если прямо не указано иного). Вам необходимо только дополнить предложенный файл.
Выполняйте каждое задание в отдельной ячейке, выделенной под него (в шаблоне они помечены как «ваш код здесь»). Не создавайте множество дополнительных ячеек — они делают ноутбук перегруженным и трудночитаемым.
В решении должен быть использован только уже пройденный материал, кроме тех случаев, когда на использование дополнительного материала будет указано отдельно. Также в кейсе вам будет предложено творческое задание (доработка модели с использованием дополнительных инструментов). В нём вы сможете использовать любой доступный материал. Это будет указано в задании явным образом.
Код должен быть читабельным и понятным: имена переменных и функций должны отражать их сущность, приветствуется отсутствие многострочных конструкций и условий. Любую задачу можно решить множеством вариантов: постарайтесь найти самый красивый и лаконичный. 
Оформите код по стандартам PEP-8. Вы можете освежить в памяти требования стандарта в соответствующем руководстве. Также вы можете воспользоваться одним из онлайн-ресурсов, который проверяет код на соответствие стандартам (однако он зачастую бывает слишком строг).
Все визуализации необходимо выполнять в соответствии с требованиями, которые вы изучали ранее. Все диаграммы и графики должны иметь содержательные названия и подписи осей. Помните, что любой человек должен без труда и дополнительных вопросов понять, что изображено на визуализации только по имеющейся на ней информации.
Оформите выводы к графикам в формате Markdown под самим графиком в отдельной ячейке (в шаблоне они помечены как «ваши выводы здесь»). Если вы хотите красиво оформить текстовые вставки, можно воспользоваться следующим ресурсом.

---

# Модель прогнозирования эффективности таргетированной маркетинговой кампании банка

Для проведения анализа и построения модели предоставлены данные о последней маркетинговой кампании банка, нацеленной на привлечение клиентов для открытия депозита.  

**Бизнес-задача**: определение характеристик для выявления клиентов, более склонных к открытию депозита в банке, обеспечивающее повышение эффективности маркетинговых кампаний.  

**Техническая задача**: построение модели машинного обучения, прогнозирующей решение клиента об открытии депозита в рамках проведения маркетинговой кампании.  

**Основные цели**: исследование данных и выявление характерных черт потенциальных клиентов для определения целевой аудитории.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.ensemble import IsolationForest
import warnings
warnings.filterwarnings('ignore')
from sklearn.preprocessing  import LabelEncoder
from sklearn import linear_model 
from sklearn import tree 
from sklearn import ensemble 
from sklearn import metrics 
from sklearn import preprocessing 
from sklearn.model_selection import train_test_split 
from sklearn.feature_selection import SelectKBest, f_classif

import plotly.graph_objects as go
from plotly.subplots import make_subplots

## Обзор данных

Данные о клиентах банка:

- age (возраст);
- job (сфера занятости);
- marital (семейное положение);
- education (уровень образования);
- default (имеется ли просроченный кредит);
- housing (имеется ли кредит на жильё);
- loan (имеется ли кредит на личные нужды);
- balance (баланс).  

Данные, связанные с последним контактом в контексте текущей маркетинговой кампании:

- contact (тип контакта с клиентом);
- month (месяц, в котором был последний контакт);
- day (день, в который был последний контакт);
- duration (продолжительность контакта в секундах).  

Прочие признаки:

- campaign (количество контактов с этим клиентом в течение текущей кампании);
- pdays (количество пропущенных дней с момента последней маркетинговой кампании до контакта в текущей кампании);
- previous (количество контактов до текущей кампании);
- poutcome (результат предыдущей маркетинговой кампании);
- deposit (результат прошедшей маркетинговой кампании) - целевой признак.

Исходя из формулировки задачи, целевой признак имеет значения откроет клиент банка депозит в результате проведения маркетинговой кампании или нет, таким образом, решение сводится к задаче ***классификации***.

In [2]:
df_clients = pd.read_csv('./data/bank_fin.csv', sep = ';')
df_clients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11162 entries, 0 to 11161
Data columns (total 17 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   age        11162 non-null  int64 
 1   job        11162 non-null  object
 2   marital    11162 non-null  object
 3   education  11162 non-null  object
 4   default    11162 non-null  object
 5   balance    11137 non-null  object
 6   housing    11162 non-null  object
 7   loan       11162 non-null  object
 8   contact    11162 non-null  object
 9   day        11162 non-null  int64 
 10  month      11162 non-null  object
 11  duration   11162 non-null  int64 
 12  campaign   11162 non-null  int64 
 13  pdays      11162 non-null  int64 
 14  previous   11162 non-null  int64 
 15  poutcome   11162 non-null  object
 16  deposit    11162 non-null  object
dtypes: int64(6), object(11)
memory usage: 1.4+ MB


In [3]:
# Подсчет пропущенных значений
print('Количество пропусков данных по признакам:')
for col in df_clients.columns:
    count_null = df_clients[col].isnull().sum()
    if count_null > 0:
        print(f'{col}: {count_null}')

Количество пропусков данных по признакам:
balance: 25


In [4]:
# Получение уникальных значений категориальных признаков
for col in df_clients.columns:
    if not pd.api.types.is_numeric_dtype(df_clients[col].dtype):
        print(f'{col}: {df_clients[col].unique()}')

job: ['admin.' 'technician' 'services' 'management' 'retired' 'blue-collar'
 'unemployed' 'entrepreneur' 'housemaid' 'unknown' 'self-employed'
 'student']
marital: ['married' 'single' 'divorced']
education: ['secondary' 'tertiary' 'primary' 'unknown']
default: ['no' 'yes']
balance: [' 2 343,00 $ ' ' 45,00 $ ' ' 1 270,00 $ ' ... ' 1 594,00 $ '
 ' 2 593,00 $ ' '-134,00 $ ']
housing: ['yes' 'no']
loan: ['no' 'yes']
contact: ['unknown' 'cellular' 'telephone']
month: ['may' 'jun' 'jul' 'aug' 'oct' 'nov' 'dec' 'jan' 'feb' 'mar' 'apr' 'sep']
poutcome: ['unknown' 'other' 'failure' 'success']
deposit: ['yes' 'no']


В категориальных признаках пропущенные значения обозначены как 'unknown', данное значение содержат:
- job (сфера занятости);
- education (уровень образования);
- contact (тип контакта с клиентом);
- poutcome (результат предыдущей маркетинговой кампании).

Также признак balance (баланс) требует преобразования в числовой.

## Первичная обработка данных

В первичной обработки данных производится заполнение пропусков и очистка от выбросов.

Признак balance, содержащий информацию о балансе клиента, имеет тип данных object, в качестве разделителя целой и дробной частей числа использована запятая, для разделенния тысячных числа использован пробел, в конце значения через пробел указана валюта (знак **$**). Признак в своей сути является непрерывным числовым и требует соответствующих преобразований.

In [5]:
# Преобразование balance во float
df_clients['balance'] = df_clients['balance'].apply(lambda x: x.replace('$', '').replace(',', '.').replace(' ', '') \
                                                              if not pd.isna(x) else x)
df_clients['balance'] = df_clients['balance'].astype(float)

In [6]:
df_clients['balance'].describe()

count    11137.000000
mean      1529.129029
std       3228.114892
min      -6847.000000
25%        123.000000
50%        550.000000
75%       1708.000000
max      81204.000000
Name: balance, dtype: float64

Ввиду небольшого количества пропусков признака balance (25 из 11162) целесообразно их заполнить медианным значением.

In [7]:
# Обработка пропусков balance
median_balance = df_clients['balance'].median()
df_clients['balance'].fillna(median_balance, inplace=True)
df_clients['balance'].describe()

count    11162.000000
mean      1526.936033
std       3224.829691
min      -6847.000000
25%        123.000000
50%        550.000000
75%       1699.000000
max      81204.000000
Name: balance, dtype: float64

После заполнения пропусков признака balance его статистические данные существенно не изменились.

In [8]:
df_clients['job'].value_counts(normalize=True)

job
management       0.229887
blue-collar      0.174162
technician       0.163322
admin.           0.119513
services         0.082691
retired          0.069701
self-employed    0.036284
student          0.032252
unemployed       0.031984
entrepreneur     0.029385
housemaid        0.024548
unknown          0.006271
Name: proportion, dtype: float64

In [9]:
df_clients['education'].value_counts(normalize=True)

education
secondary    0.490593
tertiary     0.330496
primary      0.134385
unknown      0.044526
Name: proportion, dtype: float64

Среди признаков, непосредственно характеризующих клиентов, пропуски отмечены в job (сфера деятельности) и education (образование), они составляют 0.6% и 4.5% соответственно, что дает основание для их заполнения модальными значениями.

In [10]:
# Обработка пропусков job и education
mode_job = df_clients['job'].mode()[0]
mode_education = df_clients['education'].mode()[0]
df_clients['job'].replace('unknown', mode_job, inplace=True)
df_clients['education'].replace('unknown', mode_education, inplace=True)

In [11]:
print(f"{df_clients[(df_clients['job'] == mode_job) & (df_clients['education'] == mode_education)]['balance'].mean():.2f}")

1598.88


Среднее значение баланса клиентов с модальными значениями образования и сферы деятельности (secondary и management соответственно) составляет 1 598.88 $.

Для оценки наличия выбросов в признаке balance используем график.

In [12]:
# Распределение баланса клиентов
# Инициализация полотна на два графика с общей осью X
fig_balance = make_subplots(rows=2, cols=1,
                            shared_xaxes=True, # Общая ось Х
                            x_title='Баланс, $',
                            y_title='Количество',
                            row_heights=[0.15, 0.85], # Относительная высота полотен
                            subplot_titles=("Распределение клиентов по размеру счета", ""),
                            vertical_spacing=0.02) # Высота разделителя
# Построение коробчатой диаграммы
fig_balance.add_trace(go.Box(x=df_clients['balance'],
                             marker_color = '#00B945', # Цвет маркера
                             jitter=0.5, # Разнос точек выбросов по вертикали
                             name='',
                             showlegend=False),
                      row=1, col=1)
# Построение гистограммы
fig_balance.add_trace(go.Histogram(x=df_clients['balance'],
                                   marker_color = '#00B945', # Цвет маркера
                                   name='Баланс', # Название набора данных
                                   nbinsx=100,
                                   showlegend=False),
                      row=2, col=1)
# Размер полотна
fig_balance.update_layout(autosize = False, width = 1000, height = 620,
                          bargap=0.1) # Расстояние между столбцами гистограммы
fig_balance.show()

Вид коробчатой диаграммы показывает большое количество выбросов, особенно в сторону увеличения размера баланса. Очистку от выбросов произведем с использованием метода Тьюки.

In [13]:
# Очистка данных по методу Тьюки
q25 = df_clients['balance'].quantile(0.25)
q75 = df_clients['balance'].quantile(0.75)
irq = q75 - q25
bound_upper = q75 + 1.5*irq
bound_lower = q25 - 1.5*irq
# Очистка от выбросов
df_clients = df_clients[(bound_lower <= df_clients['balance']) & (df_clients['balance'] <= bound_upper)]
print(f'Количество записей после очистки: {df_clients.shape[0]}')

Количество записей после очистки: 10105


In [14]:
# Инициализация полотна
fig_balance = go.Figure()
# Построение коробчатой диаграммы
fig_balance.add_trace(go.Histogram(x=df_clients['balance'],
                                   marker_color = '#00B945', # Цвет маркера
                                   name='Баланс', # Название набора данных
                                   nbinsx=100,
                                   showlegend=False)) # Название набора данных
fig_balance.update_layout(autosize = False, width = 1000, height = 620, # Размер полотна
                          bargap=0.1,
                          title='Распределение клиентов банка по размеру счета',
                          title_x=0.5) # Выравнивание заголовка посередине
# Подпись осей
fig_balance.update_xaxes(title_text="Баланс, $")
fig_balance.update_yaxes(title_text="Количество")
fig_balance.show()

После очистки датасета от выбросов по признаку balance (баланс клиента) осталось 10105 записей.

## Часть 2:  Разведывательный анализ

### Задание 1

In [15]:
# изучите соотношение классов в ваших данных на предмет несбалансированности, проиллюстрируйте результат
# Построение баров распределения целевого признака
fig_target = go.Figure(data=go.Bar(x=df_clients['deposit'].value_counts().index.astype(str),
                                y=df_clients['deposit'].value_counts()))
# Размер полотна
fig_target.update_layout(autosize = False, width = 850, height = 400)
# Заголовок диаграммы
fig_target.update_layout(title=f"Распределение целевого признака <br>\"deposit\"",
                        title_x=0.5, # Расположение посередине
                        xaxis_title=col, # Подпись Х
                        yaxis_title='Количество') # Подпись Y
fig_target.show()

print('Статистические данные классов целевого признака:')
display(df_clients['deposit'].value_counts(normalize=True).head(5).round(4)*100)

Статистические данные классов целевого признака:


deposit
no     53.68
yes    46.32
Name: proportion, dtype: float64

Классы целевого признака распределены следующим образом:
- открывшие депозит - 46.33 %;
- отказавшиеся - 53.67 %.  
Данные значения позволяют определить выборку как **сбалансированную**.

### Задания 2 и 3

In [16]:
# Расшифровка признаков для графиков
dict_col_name_rus = {'work_year': 'Год выплаты зарплаты',
                     'experience_level': 'Опыт работы на должности в течение года',
                     'employment_type': 'Тип трудоустройства',
                     'job_title': 'Роль, в которой соискатель работал в течение года',
                     'salary_in_usd': 'Зарплата в долларах США',
                     'employee_residence': 'Основная страна проживания сотрудника',
                     'remote_ratio': 'Общий объём работы, выполняемой удалённо',
                     'company_location': 'Страна главного офиса работодателя или филиала по контракту',
                     'company_size': 'Среднее количество людей, работавших в компании в течение года'}

In [17]:
#рассчитайте описательные статистики для количественных переменных, проинтерпретируйте результат
# Построение гистограммы распределения каждого числового признака
for col in df_clients.columns:
    if pd.api.types.is_numeric_dtype(df_clients[col].dtype):
        # Инициализация полотна на два графика с общей осью X
        fig_col = make_subplots(rows=2, cols=1,
                                shared_xaxes=True, # Общая ось Х
                                x_title=dict_col_name_rus.get(col, col),
                                y_title='Количество',
                                row_heights=[0.15, 0.85], # Относительная высота полотен
                                subplot_titles=(f"Распределение признака <br>\"{dict_col_name_rus.get(col, col)}\"", ""),
                                vertical_spacing=0.02) # Высота разделителя
        # Построение коробчатой диаграммы
        fig_col.add_trace(go.Box(x=df_clients[col],
                                 marker_color = '#1729B0', # Цвет маркера
                                 jitter=0.5, # Разнос точек выбросов по вертикали
                                 name='',
                                 showlegend=False),
                          row=1, col=1)
        # Построение гистограммы
        fig_col.add_trace(go.Histogram(x=df_clients[col],
                                       marker_color = '#1729B0', # Цвет маркера
                                       name=dict_col_name_rus.get(col, col), # Название набора данных
                                       showlegend=False),
                          row=2, col=1)
        # Построение линии среднего значения
        fig_col.add_vline(x=df_clients[col].mean(),
                          line=dict(color="Red", dash='dash')) # Стиль линии
        # Размер полотна
        fig_col.update_layout(autosize = False, width = 850, height = 550,
                              bargap=0.1) # Расстояние между столбцами гистограммы
        fig_col.show()
        
        print('Статистические данные признака:')
        display(df_clients[col].describe().round(2))

Статистические данные признака:


count    10105.00
mean        40.90
std         11.73
min         18.00
25%         32.00
50%         38.00
75%         48.00
max         95.00
Name: age, dtype: float64

Статистические данные признака:


count    10105.00
mean       807.65
std        994.15
min      -2049.00
25%         95.00
50%        445.00
75%       1227.00
max       4063.00
Name: balance, dtype: float64

Статистические данные признака:


count    10105.00
mean        15.59
std          8.44
min          1.00
25%          8.00
50%         15.00
75%         22.00
max         31.00
Name: day, dtype: float64

Статистические данные признака:


count    10105.00
mean       368.74
std        346.65
min          2.00
25%        137.00
50%        252.00
75%        490.00
max       3881.00
Name: duration, dtype: float64

Статистические данные признака:


count    10105.00
mean         2.52
std          2.71
min          1.00
25%          1.00
50%          2.00
75%          3.00
max         43.00
Name: campaign, dtype: float64

Статистические данные признака:


count    10105.00
mean        51.32
std        109.64
min         -1.00
25%         -1.00
50%         -1.00
75%          2.00
max        854.00
Name: pdays, dtype: float64

Статистические данные признака:


count    10105.00
mean         0.82
std          2.24
min          0.00
25%          0.00
50%          0.00
75%          1.00
max         58.00
Name: previous, dtype: float64

Признак **'Год выплаты зарплаты'**:  
Числовой признак является дискретным, количество значений равно трем, а промежутки между ними равны, что позволяет рассматривать его как порядковый категориальный. Количество записей возрастает с увеличением года.  
При рассмотрении года выплаты зарплаты следует отметить достаточно малое количество данных за 2020 год.

Признак **'Зарплата в долларах США'**:  
Является целевым признаком в рассматриваемых вопросах. Из формы гистограммы и неравенства среднего значения зарплаты и ее медианы следует, что распределение значений отлично от нормального, а именно является правосторонним скошенным, что означает, что большая часть данных сосредоточена в нижнем диапазоне, а небольшая часть специалистов является высокооплачиваемыми. Форма коробчатой диаграммы свидетельствует о наличии выбросов, где зарплата значительно превышает медиану. Эти данные могут быть связаны с высокооплачиваемыми руководителями или специалистами.

### Задания 4 и 5

In [18]:
#рассчитайте описательные статистики для категориальных переменных, проинтерпретируйте результат
for col in df_clients.columns:
    if not pd.api.types.is_numeric_dtype(df_clients[col].dtype):
        # Инициализация полотна на два графика с общей осью X
        fig_col = go.Figure(data=go.Bar(x=df_clients[col].value_counts().index.astype(str),
                                        y=df_clients[col].value_counts()))
        # Размер полотна
        fig_col.update_layout(autosize = False, width = 850, height = 400)
        # Заголовок диаграммы
        fig_col.update_layout(title=f"Распределение признака <br>\"{dict_col_name_rus.get(col, col)}\"",
                              title_x=0.5, # Расположение посередине
                              xaxis_title=dict_col_name_rus.get(col, col), # Подпись Х
                              yaxis_title='Количество') # Подпись Y
        fig_col.show()
        
        print('Статистические данные признака:')
        display(df_clients[col].describe())
        print('Top-5 значений (%):')
        display(df_clients[col].value_counts(normalize=True).head(5).round(4)*100)
#постройте визуализации, иллюстрирующие результаты

Статистические данные признака:


count          10105
unique            11
top       management
freq            2315
Name: job, dtype: object

Top-5 значений (%):


job
management     22.91
blue-collar    17.88
technician     16.21
admin.         12.33
services        8.59
Name: proportion, dtype: float64

Статистические данные признака:


count       10105
unique          3
top       married
freq         5715
Name: marital, dtype: object

Top-5 значений (%):


marital
married     56.56
single      31.80
divorced    11.65
Name: proportion, dtype: float64

Статистические данные признака:


count         10105
unique            3
top       secondary
freq           5517
Name: education, dtype: object

Top-5 значений (%):


education
secondary    54.60
tertiary     32.05
primary      13.35
Name: proportion, dtype: float64

Статистические данные признака:


count     10105
unique        2
top          no
freq       9939
Name: default, dtype: object

Top-5 значений (%):


default
no     98.36
yes     1.64
Name: proportion, dtype: float64

Статистические данные признака:


count     10105
unique        2
top          no
freq       5243
Name: housing, dtype: object

Top-5 значений (%):


housing
no     51.89
yes    48.11
Name: proportion, dtype: float64

Статистические данные признака:


count     10105
unique        2
top          no
freq       8712
Name: loan, dtype: object

Top-5 значений (%):


loan
no     86.21
yes    13.79
Name: proportion, dtype: float64

Статистические данные признака:


count        10105
unique           3
top       cellular
freq          7283
Name: contact, dtype: object

Top-5 значений (%):


contact
cellular     72.07
unknown      21.39
telephone     6.54
Name: proportion, dtype: float64

Статистические данные признака:


count     10105
unique       12
top         may
freq       2617
Name: month, dtype: object

Top-5 значений (%):


month
may    25.90
jul    14.03
aug    13.71
jun    10.93
apr     8.21
Name: proportion, dtype: float64

Статистические данные признака:


count       10105
unique          4
top       unknown
freq         7570
Name: poutcome, dtype: object

Top-5 значений (%):


poutcome
unknown    74.91
failure    10.97
success     9.35
other       4.76
Name: proportion, dtype: float64

Статистические данные признака:


count     10105
unique        2
top          no
freq       5424
Name: deposit, dtype: object

Top-5 значений (%):


deposit
no     53.68
yes    46.32
Name: proportion, dtype: float64

Признак **'Опыт работы на должности в течение года'**:  
Преобладает категория SE (Senior-level/Expert), за ней следует MI (Mid-level/Intermediate) и EN (Entry-level/Junior), наименьшее количество в категории EX (Executive-level/Director).  
Опыт работы на должности в течение года должен быть значимым для анализа зарплат.

Признак **'Тип трудоустройства'**:  
Явно преобладает категория FT (полный рабочий день), категории PT (неполный рабочий день) и FL (фриланс) крайне малы и сопоставимы по своему количеству между собой.  
Тип трудоустройства должен быть значимым при анализе заработной платы, однако, в данной выборке категория FT встречается в более чем 97% записей, записей по остальным категориям достаточно мало, поэтому выводы по ним могут быть недостоверны.

Признак **'Роль, в которой соискатель работал в течение года'**:  
По данному признаку была произведена фильтрация данных из исходного набора. В текущем состоянии признак содержит две категории: Data Scientist и Data Engineer, количество записей в Data Scientist преобладает незначительно, соответсвенно категории сопоставимы между собой, что позволяет провести корректный анализ по ключевым вопросам HR-агентства, в которых фигурирует сравнение показателей специалистов Data Scientist и Data Engineer.

Признак **'Основная страна проживания сотрудника'**:  
Наибольшая доля приходится на US (США) и GB (Великобритания) со значительным преобладанием US. Остальные 34% записей распределены в другой 31 категории признака.  
Страна проживания сотрудника может оказывать значительное влияние на размер его заработной платы.

Признак **'Общий объём работы, выполняемой удалённо'**:  
Преобладает категория 100 (полностью удалённая работа), за ней следует 0 (удалённой работы нет), наименьшее количество в категории 50 (частично удалённая работа).  
Общий объём работы, выполняемой удалённо, должен быть значимым для анализа зарплат.

Признак **'Страна главного офиса работодателя или филиала по контракту'**:  
Наибольшая доля приходится на US (США) и GB (Великобритания) со значительным преобладанием US. Остальные 30% записей распределены в других 27 категориях признака.  
Страна главного офиса работодателя может оказывать значительное влияние на размер его заработной платы.

Признак **'Среднее количество людей, работавших в компании в течение года'**:  
Преобладает категория M (средняя компания), за ней следует L (крупная компания), наименьшее количество в категории S (небольшая компания).  
Размер компании должен быть значимым для анализа зарплат.

### Задание 6

In [19]:
# Узнайте, для какого статуса предыдущей маркетинговой кампании успех в текущей превалирует над количеством неудач.
figure = go.Figure()
figure.add_trace(go.Bar(x=df_clients[df_clients['deposit'] == 'yes']['poutcome'].value_counts().index.astype(str),
                        y=df_clients[df_clients['deposit'] == 'yes']['poutcome'].value_counts(),
                        name='Депозит открыт',
                        marker_color = 'Green'))
figure.add_trace(go.Bar(x=df_clients[df_clients['deposit'] == 'no']['poutcome'].value_counts().index.astype(str),
                        y=df_clients[df_clients['deposit'] == 'no']['poutcome'].value_counts(),
                        name='Отказ',
                        marker_color = 'Red'))
# Размер полотна
figure.update_layout(autosize = False, width = 850, height = 400)
# Заголовок диаграммы
figure.update_layout(title=f"Распределение количества клиентов, <br>отказавшихся от открытия депозита, по месяцам",
                     title_x=0.5, # Расположение посередине
                     xaxis_title='Месяц', # Подпись Х
                     yaxis_title='Количество') # Подпись Y
figure.show()

### Задание 7

In [20]:
# узнайте, в каком месяце чаще всего отказывались от предложения открыть депозит
figure = go.Figure(data=go.Bar(x=df_clients[df_clients['deposit'] == 'no']['month'].value_counts().index.astype(str),
                               y=df_clients[df_clients['deposit'] == 'no']['month'].value_counts()))
# Размер полотна
figure.update_layout(autosize = False, width = 850, height = 400)
# Заголовок диаграммы
figure.update_layout(title=f"Распределение количества клиентов, <br>отказавшихся от открытия депозита, по месяцам",
                     title_x=0.5, # Расположение посередине
                     xaxis_title='Месяц', # Подпись Х
                     yaxis_title='Количество') # Подпись Y
figure.show()

Чаще всего отказывались открыть депозит в мае (1775 клиентов).

### Задание 8

В соответствии со справочной информацией клиенты подразделяются на следующие возрастные категории:
- менее 22 лет (молодежь)
- 22-34 лет (молодые люди, образовавшие семью)
- 35-54 лет (лица "зрелого возраста")
- 55 и более лет (лица, готовящиеся к уходу на пенсию)

Распределение на категории по данным признакам вносится в новые: **CreditScoreCat** и **AgeCat** соответственно.

In [21]:
# создайте возрастные группы и определите, в каких группах более склонны открывать депозит, чем отказываться от предложения
# Функция распределения возраста по категориям
def get_age_cat(age):
    if age >= 30 and age < 40:
        return "30-40"
    elif age >= 40 and age < 50:
        return "40-50"
    elif age >= 50 and age < 60:
        return "50-60"
    elif age >= 60:
        return "60+"
    elif age < 30:
        return "<30" 


# Создание признака категорий возраста
df_clients['age_cat'] = df_clients['age'].apply(get_age_cat)

In [22]:
# Инициализация полотна на три графика
figure = go.Figure()
figure.add_trace(go.Bar(x=df_clients[df_clients['deposit'] == 'yes']['age_cat'].value_counts().index.astype(str),
                        y=df_clients[df_clients['deposit'] == 'yes']['age_cat'].value_counts(),
                        name='Депозит открыт',
                        marker_color = 'Green'))
figure.add_trace(go.Bar(x=df_clients[df_clients['deposit'] == 'no']['age_cat'].value_counts().index.astype(str),
                        y=df_clients[df_clients['deposit'] == 'no']['age_cat'].value_counts(),
                        name='Отказ',
                        marker_color = 'Red'))
# Размер полотна
figure.update_layout(autosize = False, width = 850, height = 400)
# Порядок вывода категорий
figure.update_xaxes(categoryorder='array',
                    categoryarray=["<30", "30-40", "40-50", "50-60", "60+"]) 
# Заголовок диаграммы
figure.update_layout(title=f"Распределение количества клиентов, <br>открывших и неоткрывших депозит, по возрастным категориям",
                     title_x=0.5, # Расположение посередине
                     xaxis_title='Возрастная категория', # Подпись Х
                     yaxis_title='Количество') # Подпись Y         
figure.show()

Более склонны открывать депозит, чем отказываться от него, в возрастных категориях "55 лет и старше" и "менее 22 лет", при этом наибольшее количество открытия депозитов совершили клиенты возрастной категории 35-54, но и максимальное количество отказов принадлежит данной группе, причем в полтора раза больше, чем открытий.

### Задания 9 и 10

In [23]:
# постройте визуализации для открывших и неоткрывших депозит в зависимости от семейного статуса
figure = go.Figure()
figure.add_trace(go.Bar(x=df_clients[df_clients['deposit'] == 'yes']['marital'].value_counts().index.astype(str),
                        y=df_clients[df_clients['deposit'] == 'yes']['marital'].value_counts(),
                        name='Депозит открыт',
                        marker_color = 'Green'))
figure.add_trace(go.Bar(x=df_clients[df_clients['deposit'] == 'no']['marital'].value_counts().index.astype(str),
                        y=df_clients[df_clients['deposit'] == 'no']['marital'].value_counts(),
                        name='Отказ',
                        marker_color = 'Red'))
# Размер полотна
figure.update_layout(autosize = False, width = 850, height = 400)
# Заголовок диаграммы
figure.update_layout(title=f"Распределение количества клиентов, <br>открывших и неоткрывших депозит, по семейному положению",
                     title_x=0.5, # Расположение посередине
                     xaxis_title='Семейное положение', # Подпись Х
                     yaxis_title='Количество') # Подпись Y         
figure.show()

In [24]:
# постройте визуализации для открывших и неоткрывших депозит в зависимости от образования
figure = go.Figure()
figure.add_trace(go.Bar(x=df_clients[df_clients['deposit'] == 'yes']['education'].value_counts().index.astype(str),
                        y=df_clients[df_clients['deposit'] == 'yes']['education'].value_counts(),
                        name='Депозит открыт',
                        marker_color = 'Green'))
figure.add_trace(go.Bar(x=df_clients[df_clients['deposit'] == 'no']['education'].value_counts().index.astype(str),
                        y=df_clients[df_clients['deposit'] == 'no']['education'].value_counts(),
                        name='Отказ',
                        marker_color = 'Red'))
# Размер полотна
figure.update_layout(autosize = False, width = 850, height = 400)
# Заголовок диаграммы
figure.update_layout(title=f"Распределение количества клиентов, <br>открывших и неоткрывших депозит, по уровню образования",
                     title_x=0.5, # Расположение посередине
                     xaxis_title='Уровень образования', # Подпись Х
                     yaxis_title='Количество') # Подпись Y         
figure.show()

In [25]:
# постройте визуализации для открывших и неоткрывших депозит в зависимости от вида профессиональной занятости
figure = go.Figure()
figure.add_trace(go.Bar(x=df_clients[df_clients['deposit'] == 'yes']['job'].value_counts().index.astype(str),
                        y=df_clients[df_clients['deposit'] == 'yes']['job'].value_counts(),
                        name='Депозит открыт',
                        marker_color = 'Green'))
figure.add_trace(go.Bar(x=df_clients[df_clients['deposit'] == 'no']['job'].value_counts().index.astype(str),
                        y=df_clients[df_clients['deposit'] == 'no']['job'].value_counts(),
                        name='Отказ',
                        marker_color = 'Red'))
# Размер полотна
figure.update_layout(autosize = False, width = 850, height = 400)
# Заголовок диаграммы
figure.update_layout(title=f"Распределение количества клиентов, <br>открывших и неоткрывших депозит, по виду профессиональной занятости",
                     title_x=0.5, # Расположение посередине
                     xaxis_title='Вид профессиональной занятости', # Подпись Х
                     yaxis_title='Количество') # Подпись Y         
figure.show()

### Задание 11

In [26]:
# постройте сводную таблицу, чтобы определить люди с каким образованием и семейным статусом наиболее многочисленны
#(если рассматривать тех, кто открыл депозит)
# Формирование сводной таблицы
pt_education_marital = df_clients[df_clients['deposit'] == 'yes'].pivot_table(values='deposit',
                                                                              index='education',
                                                                              columns='marital',
                                                                              aggfunc='count',
                                                                              observed=True) # Использовать все категории
display(pt_education_marital)
# Построение тепловой карты
figure = go.Figure(data=go.Heatmap(x=pt_education_marital.columns,
                                   y=pt_education_marital.index[::-1],
                                   z=pt_education_marital.values[::-1],
                                   colorscale='Greens',
                                   colorbar_title = 'Количество'))
# Размер полотна
figure.update_layout(autosize = False, width = 1000, height = 520)
# Заголовк и подписи осей
figure.update_layout(title='Количество клиентов, открывших депозит,\
<br>в зависимости от уровня образования и семейного положения',
                     title_x=0.5, # Выравнивание заголовка посередине
                     yaxis_title='Уровень образования', # Подпись Y
                     xaxis_title='Семейное положение') # Подпись Х
figure.show()

marital    divorced  married  single
education                           
primary          96      339      82
secondary       283     1289     879
tertiary        174      784     755

## Часть 3: преобразование данных

### Задание 1

In [ ]:
# преобразуйте уровни образования

### Задания 2 и 3

In [ ]:
# преобразуйте бинарные переменные в представление из нулей и единиц

### Задание 4

In [ ]:
# создайте дамми-переменные

### Задания 5 и 6

In [ ]:
# постройте корреляционную матрицу и оцените данные на предмет наличия мультиколлинеарности

### Задания 7 и 8

In [ ]:
X = df_clients.drop(['deposit'], axis=1)
y = df_clients['deposit']
 
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, random_state = 42, test_size = 0.33)

In [ ]:
# рассчитайте необходимые показатели

### Задание 9

In [ ]:
# с помощью SelectKBest отберите 15 наиболее подходящих признаков

### Задание 10

In [ ]:
# нормализуйте данные с помощью minmaxsxaler

# Часть 4: Решение задачи классификации: логистическая регрессия и решающие деревья

### Задание 1

In [ ]:
# обучите логистическую регрессию и рассчитайте метрики качества

### Задания 2,3,4

In [ ]:
# обучите решающие деревья, настройте максимальную глубину

### Задание 5

In [ ]:
# подберите оптимальные параметры с помощью gridsearch

# Часть 5: Решение задачи классификации: ансамбли моделей и построение прогноза

### Задание 1

In [ ]:
# обучите на ваших данных случайный лес

### Задания 2 и 3

In [ ]:
# используйте для классификации градиентный бустинг и сравните качество со случайным лесом

### Задание 4

In [ ]:
# объедините уже известные вам алгоритмы с помощью стекинга 

### Задание 5

In [ ]:
# оцените, какие признаки демонстрируют наибольшую  важность в модели градиентного бустинга

### Задания 6,7,8

In [ ]:
# реализуйте оптимизацию гиперпараметров с помощью Optuna